climada/engine/impact.py

```diff
@@ -1372,7 +1372,147 @@ def _selected_exposures_idx(self, coord_exp):
+    def concat(cls, imp_list,concat_type='time_window'):
+        """Concatenate impact objects of the same hazard type, frequency, 
+        exposure, and coordinate system.
 
+        Parameters
+        ----------
+        imp_list : list of climada.engine.Impact objects
+                list of Impact object to concatenate
+        concat_type: string
+                string specifying the type of concatenation to be done.
+                    'time_window': concatenate impacts for different time windows / time slices within a year (e.g. May to June, July to August,...)
+                    'exposure': concatenate impacts with separate exposures
+        Returns
+        --------
+        impact: climada.engine.Impact
+               New impact object which is a concatenation of all impacts
+        """
+       
+        if len(imp_list) == 0:
+           return cls()
+       
+        if concat_type not in ['time_window','exposure']:
+           raise ValueError(f"Concatenation option {concat_type} not available.")
+       
+        #initialize new impact object
+        impact=cls()
+       
+        # check impacts consistency and fill impact object
+       
+        ### the following checks are common for both concat options
+       
+        # check hazard type
+        haz_types = {imp.tag['haz'].haz_type for imp in imp_list if imp.tag['haz'].haz_type != ''}
+        if len(haz_types) > 1:
+           raise ValueError(f"The impacts are based on hazards of different types: {haz_types}. "
+                            "The impacts are incompatible and cannot be concatenated.")      
+       
+        # check unit
+        units = {imp.unit for imp in imp_list if imp.unit != ''}
+        if len(units) > 1:
+           raise ValueError(f"The given impacts have different units: {units}. "
+                            "The impacts are incompatible and cannot be concatenated.")
+        if len(units) == 0:
+           units = {''}
+        impact.unit = units.pop()
+
+        #check coordinate reference system
+        crss={imp.crs for imp in imp_list if imp.crs != ''}
+        if len(crss)>1:
+           raise ValueError(f"The given impacts have different crs: {crss}. "
+                        "The impacts are incompatible and cannot be concatenated.")
+        impact.crs=crss.pop()
+ 
+        #check attributes
+        attributes = sorted(set.union(*[set(vars(imp).keys()) for imp in imp_list]))
+        for attr_name in attributes:
+           if not all(hasattr(imp, attr_name) for imp in imp_list):
+               raise ValueError(f"Attribute {attr_name} is not shared by all impacts. "
+                                "The impacts are incompatible and cannot be concatenated.")
+
+        #check if impacts have same frequency values (only first entries are checked)
+        freqs=set([imp.frequency[0] for imp in imp_list])
+        if len(freqs)>1:
+           raise ValueError(f"Impacts do not have the same frequency: {freqs}."
```

Comment:
Hmm...It does work for time_window if the impacts are computed from the same original hazard set, right? Because then they all have the same frequency. I would argue that this is needed for concatenation because adding impacts with different frequencies does not seem to be reasonable if we don't include some kind of weighting. And numpy.array_equal does not work because impacts can have a different number of events.

Reply:
Events in the same hazard in general do not have to have the same frequencies.

Reply:
ok then I have to think about if in such a case all the rest of the concatenation is still correct...